backend/core/src/csv/csv-builder.service.ts

```diff
@@ -1,130 +1,330 @@
+  }
+
+  private addToExlucdedKeys(key) {
+    this.excludeKeys.push(key)
+  }
+
+  private parseApplication(key: string, val: any, header: string, index = 0): void {
+    if (Array.isArray(val)) {
+      val.forEach((val2, i) => {
+        // preferences are a special case
+        if (key === "preferences") {
+          val2.options?.forEach((preference) => {
+            this.parseApplication(
+              preference.key,
+              preference.checked ? "Yes" : "No",
+              `${header} ${val2.key} ${preference.key}`
+            )
+          })
+        } else {
+          this.parseApplication(key, val2, header, i + 1)
         }
+      })
+    } else if (val instanceof Object) {
+      // dates are objects
+      if (dayjs(val).isValid()) {
+        const [newKey, mappedField] = this.setHeader(header, index)
+        this.data.push({
+          header: newKey,
+          mappedField,
+          val: dayjs(val).format("DD-MM-YYYY h:mm:ss A"),
+        })
       } else {
-        headers.push(metadata.label)
+        for (let i = 0, keys = Object.keys(val); i < keys.length; i++) {
+          this.parseApplication(
+            keys[i],
+            val[keys[i]],
+            header.length ? `${header} ${keys[i]}` : keys[i],
+            index
+          )
+        }
       }
+    } else {
+      const [newKey, mappedField] = this.setHeader(header, index)
+      // format data before pushing
+      if (typeof val === "boolean") {
+        val = val ? "Yes" : "No"
+      }
+      // use JSON.stringify to escape double and single quotes
+      this.data.push({
+        header: newKey,
+        mappedField,
+        val: val !== undefined && val !== null ? JSON.stringify(val) : "",
+      })
     }
-    return headers
   }
 
-  private normalizeMetadataArrays(
-    arr: any[],
-    formattingMetadataAggregate: FormattingMetadataAggregate
-  ) {
-    formattingMetadataAggregate
-      .filter((metadata) => metadata.type === "array")
-      .forEach((metadata) => {
-        const md = metadata as FormattingMetadataArray
-        if (md.size !== null) {
-          return
-        }
-        md.size = Math.max(
-          ...arr.map((item) => {
-            const value = this.retrieveValueByDiscriminator(item, md.discriminator)
-            if (!value || !Array.isArray(value)) {
-              return 0
+  public build(arr: any[], includeDemographics?: boolean): string {
+    if (this.inProgress === true) {
+      throw new HttpException(
+        {
+          status: HttpStatus.TOO_MANY_REQUESTS,
+          error: "Download Already in Progress",
+        },
+        HttpStatus.TOO_MANY_REQUESTS
+      )
+    }
+    this.setProgress(true)
+    let dataString = ""
+    try {
+      if (!includeDemographics) {
+        this.addToExlucdedKeys("demographics")
+      }
+      arr.forEach((row: ApplicationDto) => {
+        this.parseApplication("", row, "")
+        this.data.push({ header: "_endRow", mappedField: undefined, val: "_endRow" })
```

Comment:
What the purpose of _endRow?

Reply:
_endRow is used below to determine when we're at the end of a row and to go to the next one. Farther down `rows` is defined as an array of arrays with all of the necessary slots created, according to the header columns and the length of `arr`, so then we can fill data rows according to their matching header.